### PR TITLE
Disk spilling support for hash aggregations

### LINKED ISSUE #9609: colexec: add the disk-spilling to the hash aggregator
PR: https://github.com/cockroachdb/cockroach/pull/57817

From release notes:
> Hash aggregation can now spill to disk when it exhausts its memory limit when executed via the [vectorized engine](../v21.1/vectorized-execution.html). [#57817][#57817] {% comment %}doc{% endcomment %}

### LINKED ISSUE #9922: sql: document known limitations around disk spilling for unordered distinct and hash aggregation
Up until and including 20.2 version, we didn't support spilling to disk for unordered distinct and hash aggregation operations. The behavior was that we would buffer all of the relevant data in RAM (thus, possibly exceeding `sql.distsql.temp_storage.workmem` limit) until the query hits the node-wide `--max-sql-memory` limit, at which point it'll error out. Note that it is applicable to both row-by-row and vectorized execution engines. Furthermore, our memory accounting is imprecise (especially in the row-by-row engine), so it is possible that instead of erroring out with a memory limit, the node will crash.

In 21.1 version we merged the full support for disk spilling for hash aggregation and general case support for unordered distinct in the vectorized engine (the row-by-row engine will be deprecated, so we don't invest effort there). The missing support for unordered distinct is when it is used for UPSERT queries and has "nulls are distinct" and "error on duplicate" markers mentioned in EXPLAIN:
```
        ├── distinct                     |                     |
        │    │                           | distinct on         | ...
        │    │                           | nulls are distinct  |
        │    │                           | error on duplicate  |
```
The tracking issue for this case is https://github.com/cockroachdb/cockroach/issues/61411.

### PR DESCRIPTION
**Pull request by ianjevans**
Fixes #9609
Fixes #9922

Also clarifies the lack of support for disk spilling for hash aggregations in 20.2 docs.

### CODE REVIEW

File: v21.1/vectorized-execution.md [hunk 1]
```diff
@@ -51,6 +51,7 @@ The following operations require [memory buffering](https://en.wikipedia.org/wik
 
 - Global [sorts](order-by.html)
 - [Unordered aggregations](order-by.html)
+- Hash aggregations
```
Comment (yuzefovich) [inline v21.1/vectorized-execution.md]: "Hash" and "unordered" aggregations are synonyms.

**Review by yuzefovich — COMMENTED** [review]
The wording sounds good to me.

Oh, I see - you have two commits - you should squash them into one.

Also, there is this issue that is very related to this PR: https://github.com/cockroachdb/docs/issues/9922.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338#-:-MYHSjAT6ywWtyv6X94_:b-kh50bj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ianjevans)

___
*[v21.1/vectorized-execution.md, line 58 at r1](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYHS5sB9RgVtM9o3OzQ:-MYHS5sB9RgVtM9o3OzR:bi4mzc3) ([raw file](https://github.com/cockroachdb/docs/blob/c67d3988a76227226780f3a7142311687f7fa0b8/v21.1/vectorized-execution.md#L58)):*
> ```Markdown
> - [Window functions](window-functions.html). Note that [support for window functions is limited in the vectorized execution engine](#window-functions).
> 
> If there is not enough memory allocated for an operation, CockroachDB will spill the intermediate execution results to disk. By default, the memory limit allocated per operator is 64MiB. You can change this limit with the `sql.distsql.temp_storage.workmem` [cluster setting](cluster-settings.html).
> ```

Actually MiB is correct here - we use 64 * 2 ^ 10 internally. (I do sometimes say 64MB out of laziness :) ).

**Review by yuzefovich — APPROVED** [review]
Two commits should be squashed, but otherwise :lgtm: thanks!

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338#-:-MYHe2sNDp1JRE2WqGKs:bc4699q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ianjevans)

___
*[v21.1/vectorized-execution.md, line 78 at r2](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYHdqU628v6w0kVMRhp:-MYHdqU71IAdZYZKOdYM:b-ekbytr) ([raw file](https://github.com/cockroachdb/docs/blob/9c6d1991a87f6a1290d712e850ef84d4336b8d38/v21.1/vectorized-execution.md#L78)):*
> ```Markdown
> The vectorized engine does not support [working with spatial data](spatial-data.html). Queries with [geospatial functions](functions-and-operators.html#spatial-functions) or [spatial data](spatial-data.html) will revert to the row-oriented execution engine.
> 
> ### Unordered distinct aggregations
> ```

nit: I'd replace "aggregations" with "operations".

**Comment by ianjevans** [comment]
TFTR, @yuzefovich!

The commits will be squashed on merged.

**Review by ericharmeling — COMMENTED** [review]
Looks good! I had a few comments on known limitations docs. See below.

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338#-:-MYL9sxI7h5TZSkgLpsg:b-ywwevk)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @awoods187, @ianjevans, and @kevin-v-ngo)

___
*[v20.2/vectorized-execution.md, line 64 at r3](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYL9Wxq0uhMHNYB3Tfi:-MYL9Wxq0uhMHNYB3Tfj:b-sujlyu) ([raw file](https://github.com/cockroachdb/docs/blob/e3c36501bccff5637e424fb253241bf2bc0ebab4/v20.2/vectorized-execution.md#L64)):*
> ```Markdown
> 
> ## Known limitations
> ```

Are unordered distinct operations not also a limitation in 20.2?
___
*[v20.2/vectorized-execution.md, line 65 at r3](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYL7ZsFE2UO7oRLtOj2:-MYL7ZsFE2UO7oRLtOj3:b-bdi6xq) ([raw file](https://github.com/cockroachdb/docs/blob/e3c36501bccff5637e424fb253241bf2bc0ebab4/v20.2/vectorized-execution.md#L65)):*
> ```
>  Unordered aggregations
> ```

See my comment on `v21.1/vectorized-execution.md` re: known limitations. IDK if there is/was a separate GH issue that tracked unordered hash aggregations, but if there was, then we missed it last release and should probably link to that issue, following the known limitations conventions.
___
*[v21.1/vectorized-execution.md, line 78 at r3](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYL9Eod0dQJkqmJ89EM:-MYL9Eod0dQJkqmJ89EN:b-s9brcx) ([raw file](https://github.com/cockroachdb/docs/blob/e3c36501bccff5637e424fb253241bf2bc0ebab4/v21.1/vectorized-execution.md#L78)):*
> ```
> Unordered distinct operations
> ```

As Jesse noted here (https://github.com/cockroachdb/docs/issues/9922#issuecomment-790026684), this limitation might also need to be added to `known-limitations.html`. The preferred convention is making an include file with the limitation description, with a link to the `cockroach` GH issue tracking the limitation (https://github.com/cockroachdb/docs/issues/9922), and then including that include file in both the feature-specific page (`vectorized-execution.html`) and the `known-limitations` page. There are plenty of precedents for this in `_includes/v21.1/known-limitations`.

You'll note that https://github.com/cockroachdb/cockroach/issues/61411 has the label "docs-known-limitation". At the end of the release cycle, we go through and make sure that all of the issues with this label are properly documented on the `known-limitations.html` page.

**Comment by ianjevans** [comment]
> Are unordered distinct operations not also a limitation in 20.2?

Unordered distinct operations are a subset of unordered aggregation operations, all of which can't be spilled to disk in 20.2. I'll make this consistent, though, to clarify.

> > ```
> >  Unordered aggregations
> > ```
> 
> See my comment on `v21.1/vectorized-execution.md` re: known limitations. IDK if there is/was a separate GH issue that tracked unordered hash aggregations, but if there was, then we missed it last release and should probably link to that issue, following the known limitations conventions.

Done.

> As Jesse noted here ([#9922 (comment)](https://github.com/cockroachdb/docs/issues/9922#issuecomment-790026684)), this limitation might also need to be added to `known-limitations.html`. The preferred convention is making an include file with the limitation description, with a link to the `cockroach` GH issue tracking the limitation (#9922), and then including that include file in both the feature-specific page (`vectorized-execution.html`) and the `known-limitations` page. There are plenty of precedents for this in `_includes/v21.1/known-limitations`.
> 
> You'll note that [cockroachdb/cockroach#61411](https://github.com/cockroachdb/cockroach/issues/61411) has the label "docs-known-limitation". At the end of the release cycle, we go through and make sure that all of the issues with this label are properly documented on the `known-limitations.html` page.

Done.

**Review by ericharmeling — APPROVED** [review]
Nice! LGTM!

> Reviewed 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338#-:-MYLJTFb6h-IDi8UxXLr:b-ohlbpw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @awoods187, @ianjevans, and @kevin-v-ngo)

**Review by awoods187 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ianjevans and @kevin-v-ngo)

___
*[_includes/v20.2/known-limitations/unordered-operations.md, line 1 at r4](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYLIEX-F8Cm6HFhl8lZ-r4-1:-MYeLq-FC-kxlcmWFUJQ:bsufhge) ([raw file](https://github.com/cockroachdb/docs/blob/ba08efbb479108ee8108e7e868ab86992fa7b351/_includes/v20.2/known-limitations/unordered-operations.md#L1)):*
<details><summary><i>Previously, kevin-v-ngo wrote…</i></summary><blockquote>

The workaround is to then increase the '--max-sql-memory' setting correct? Can we provide a link to how to do this? I found the following doc when start starting up a node: https://www.cockroachlabs.com/docs/v20.2/cockroach-start#general. (As a workaround, increase --max-sql-memory when starting your cluster)
</blockquote></details>

I like the idea of mentioning this but we should be careful in how we phrase this because upping it too much can cause a lot of other problems.

File: v20.2/vectorized-execution.md [hunk 2]
```diff
@@ -56,12 +56,16 @@ The following operations require [memory buffering](https://en.wikipedia.org/wik
 - [Merge joins](joins.html#merge-joins) on non-unique columns. Merge joins on columns that are guaranteed to have one row per value, also known as "key columns", can execute entirely in-memory.
 - [Window functions](window-functions.html). Note that [support for window functions is limited in the vectorized execution engine](#window-functions).
 
-If there is not enough memory allocated for an operation, CockroachDB will spill the intermediate execution results to disk. By default, the memory limit allocated per operator is 64MiB. You can change this limit with the `sql.distsql.temp_storage.workmem` [cluster setting](cluster-settings.html).
+If there is not enough memory allocated for an operation, CockroachDB will spill the intermediate execution results to disk if the operation supports disk spilling. By default, the memory limit allocated per operator is 64MiB. You can change this limit with the `sql.distsql.temp_storage.workmem` [cluster setting](cluster-settings.html).
```
[the remark this answers is not in the document]
Comment (ianjevans) [reply v20.2/vectorized-execution.md:59]: We pull the cluster settings from a generated file in the `cockroachdb/cockroach` repo.
https://raw.githubusercontent.com/cockroachdb/cockroach/release-21.1/docs/generated/settings/settings.html

I don't know why that setting isn't showing up, but I will follow up with the Queries team.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10338)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ericharmeling, @ianjevans, and @kevin-v-ngo)

___
*[_includes/v21.1/known-limitations/unordered-distinct-operations.md, line 1 at r4](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYLIEX-F8Cm6HFhl8l_-r4-1:-MYgBtXTDklq-uubgr52:bb77ufg) ([raw file](https://github.com/cockroachdb/docs/blob/ba08efbb479108ee8108e7e868ab86992fa7b351/_includes/v21.1/known-limitations/unordered-distinct-operations.md#L1)):*
<details><summary><i>Previously, kevin-v-ngo wrote…</i></summary><blockquote>

Is the workaround the same for unordered aggregation operations in 20.2 and to increase --max-sql-memory?
</blockquote></details>

Yes.
___
*[v20.2/vectorized-execution.md, line 64 at r3](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYL9Wxq0uhMHNYB3Tfi:-MYgC2SV34lX3I8DvQUm:b267bek) ([raw file](https://github.com/cockroachdb/docs/blob/e3c36501bccff5637e424fb253241bf2bc0ebab4/v20.2/vectorized-execution.md#L64)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

Are unordered distinct operations not also a limitation in 20.2?
</blockquote></details>

Yes, they are.
___
*[v20.2/vectorized-execution.md, line 65 at r3](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYL7ZsFE2UO7oRLtOj2:-MYgCB1dEifJoBtYLzSf:bg82sdp) ([raw file](https://github.com/cockroachdb/docs/blob/e3c36501bccff5637e424fb253241bf2bc0ebab4/v20.2/vectorized-execution.md#L65)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

> ```
>  Unordered aggregations
> ```

See my comment on `v21.1/vectorized-execution.md` re: known limitations. IDK if there is/was a separate GH issue that tracked unordered hash aggregations, but if there was, then we missed it last release and should probably link to that issue, following the known limitations conventions.
</blockquote></details>

This was the issue: https://github.com/cockroachdb/cockroach/issues/42485. It is applicable to row-by-row engine in all releases and the vectorized engine for 20.2 and prior releases.
___
*[v20.2/vectorized-execution.md, line 59 at r4](https://reviewable.io/reviews/cockroachdb/docs/10338#-MYHQ_dtBwcC-dqXEhY9-r4-59:-MYgBwf18wTc2bKSaH_o:blqbp6q) ([raw file](https://github.com/cockroachdb/docs/blob/ba08efbb479108ee8108e7e868ab86992fa7b351/v20.2/vectorized-execution.md#L59)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

We pull the cluster settings from a generated file in the `cockroachdb/cockroach` repo.
https://raw.githubusercontent.com/cockroachdb/cockroach/release-21.1/docs/generated/settings/settings.html

I don't know why that setting isn't showing up, but I will follow up with the Queries team.
</blockquote></details>

This setting is not marked as "public", so it is not being included into the generated page. I'm not sure whether we have guidance of what constitutes a "public" (i.e. documented) vs "private" (i.e. undocumented) setting. I think `sql.distsql.temp_storage.workmem` falls right on the boundary. It has been private since its introduction years ago, but we have been mentioning it here and there occasionally.

**Pull request merged**